core/src/types/list.rs

```diff
@@ -80,43 +83,87 @@ impl Stream for Lister {
+
+                return match rp {
+                    Ok(rp) => {
+                        self.task_queue.pop_back();
+                        let metadata = rp.into_metadata();
+                        Poll::Ready(Some(Ok(Entry::new(path, metadata))))
+                    }
+                    Err(err) => {
+                        self.errored = true;
+                        Poll::Ready(Some(Err(err)))
+                    }
+                };
+            }
         }
 
-        match ready!(self.lister.poll_next(cx)) {
-            Ok(Some(oe)) => {
-                let (path, metadata) = oe.into_entry().into_parts();
-                if metadata.contains_metakey(self.required_metakey) {
-                    return Poll::Ready(Some(Ok(Entry::new(path, metadata))));
-                }
-
+        if let Some(entry) = self.buf.clone() {
+            let (path, metadata) = entry.into_parts();
+            return if metadata.contains_metakey(self.required_metakey) {
+                self.buf = None;
+                Poll::Ready(Some(Ok(Entry::new(path, metadata))))
+            } else {
                 let acc = self.acc.clone();
                 let fut = async move {
                     let res = acc.stat(&path, OpStat::default()).await;
-
                     (path, res)
                 };
-                self.stating = Some(Box::pin(fut));
+
+                self.task_queue.push_front(tokio::spawn(fut));
+                self.buf = None;
                 self.poll_next(cx)
-            }
-            Ok(None) => Poll::Ready(None),
-            Err(err) => {
-                self.errored = true;
-                Poll::Ready(Some(Err(err)))
-            }
+            };
+        }
+
+        loop {
+            return match ready!(self.lister.poll_next(cx)) {
+                Ok(Some(oe)) => {
+                    let (path, metadata) = oe.into_entry().into_parts();
+                    // TODO: we can optimize this by checking the provided metakey provided by services.
+                    if metadata.contains_metakey(self.required_metakey) {
+                        return if self.task_queue.is_empty() {
+                            Poll::Ready(Some(Ok(Entry::new(path, metadata))))
+                        } else {
+                            self.buf = Some(Entry::new(path, metadata));
+                            self.poll_next(cx)
+                        };
+                    }
+
+                    if self.task_queue.len() < self.task_queue.capacity() {
+                        let acc = self.acc.clone();
+
+                        let fut = async move {
+                            let res = acc.stat(&path, OpStat::default()).await;
+                            (path, res)
+                        };
+
+                        self.task_queue.push_front(tokio::spawn(fut));
+                        continue;
+                    } else {
+                        self.buf = Some(Entry::new(path, metadata));
```

Comment:
We can pre-check the `task_queue` capacity before so that we don't need an extra `buf`.

Reply:
Commented in the code TBD.